R/control_chart.R

```diff
@@ -0,0 +1,139 @@
+#' @param measure variable of interest mapped to y-axis
+#' @param x variable to go on the x-axis, often a time variable. If unspecified
+#' row indices will be used
+#' @param group1 Optional grouping variable to be panelled horizontally
+#' @param group2 Optional grouping variable to be panelled vertically
+#' @param center_line Function used to calculate central tendency. 
+#' Defaults to mean
+#' @param sigmas Number of standard deviations above and below the central 
+#' tendency to call a point influenced by "special cause variation."
+#' Defaults to 3
+#' @param save_to Optional file path to save chart. If not provided, the chart
+#' will be printed on screen
+#' @param plot_width In inches. Only used if save_to is specified
+#' @param plot_height In inches. Only used if save_to is specified
+#' @param plot_title Title in upper-left
+#' @param plot_catpion Caption in lower-right
+#' @param plot_font_size Base font size; text elements will be scaled to this
+#'
+#' @return Generally called for the side effect of printing the control chart
+#' or writing the control chart to file, depending on whether save_to is 
+#' specified. Invisibly, returns a ggplot object for further customization.
+#' @export
+#' @import ggplot2
+#' @importFrom readr read_csv
+#' @importFrom tibble data_frame
+#' @importFrom stats as.formula
+#'
+#' @examples
+#' d <- 
+#'   tibble::data_frame(
+#'     day = sample(c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday"), 
+#'                  100, TRUE),
+#'     person = sample(c("Tom", "Jane", "Alex"), 100, TRUE),
+#'     count = rbinom(100, 20, ifelse(day == "Friday", .5, .2)),
+#'     date = Sys.Date() - sample.int(100))
+#' control_chart(d, "count")
+#' control_chart(d, "count", group1 = "day", group2 = "person")
+#' if (require(ggplot2)) {
+#' x <- control_chart(d, "count", "date")
+#' x + 
+#'   ylab("Number of Adverse Events") + 
+#'   scale_x_date(name = "Week of", date_breaks = "week") +
+#'   theme(axis.text.x = 
+#'                    element_text(angle = -90, vjust = 0.5, hjust=1))
+#' }
+control_chart <- function(d, measure, x, group1, group2,
+                          center_line = mean, sigmas = 3,
+                          save_to, plot_width = 8, plot_height = 4,
+                          plot_title = NULL, plot_catpion = NULL,
+                          plot_font_size = 11) {
+
+  theme_set(theme_bw())
+
+  if (missing(d) || !(is.data.frame(d) || is.character(d))) {
+    stop("You have to provide a data frame or a file location.")
+  } else if (is.character(d)) {
+    message("Attempting to read csv from ", d)
+    d <- readr::read_csv(d)
+  }
+
+  if (missing(measure)) {
+    stop("You have to provide a measure variable name.")
+  }
+
+  if (missing(x)) {
+    x <- "x"
+    d$x <- seq_len(nrow(d))
+  }
+
+  bounds <- calculate_bounds(d, measure, center_line, sigmas)
+
+  # Calculate central tendency and upper and lower limits
+  d$outside <- ifelse(d[[measure]] > bounds[["upper"]] |
```

Comment:
This could be factored out into another tested function just like the bounds calcs

Reply:
@Aylr I spent 20 minutes going down this road, but there are several things calculated here that the function needs further down -- the data frame, the bounds, and if argument x wasn't provided, the value "x" in its place, and passing so many things in and returning so many things back is just ugly. Maybe this is a sign that the overall organization isn't as clean as it could be, but given that this isn't core infrastructure, I don't think it's worth getting too deep into.